### DISCUSSION TITLE
Serial port redirection to stdio or a file

### DISCUSSION THREAD

**Discussion by degaart** [86Box/86Box#3575] [Ideas]
**Is your feature request related to a problem? Please describe.**
I'm currently debugging some VGA-related code inside 86box and pcem, and need a way to output debug information to somewhere else than the screen.

**Describe the solution you'd like**
Redirect serial output to host stdio or to a file.

**Describe alternatives you've considered**
IsaBugger: no scrollback.

Usb-to-serial adapters and use another PC as serial terminal: inconvenient.

**Additional context**
qemu, bochs, virtualbox and vmware can output serial to a file. Also, qemu and bochs implement a virtual device at port 0xE9 that outputs data to the console or a log file.

**Edit**: What does the ENABLE_SERIAL_LOG preprocessor symbol do? Does it redirect serial output to 86box's log? Maybe this can be converted to a command-line switch?

**Comment by cold-brewed** [comment]
While this is not natively supported, in the meantime you could likely accomplish the same thing using `socat` and serial passthrough. You would just redirect the virtual serial port to a file.

`ENABLE_*_LOG` is generally used for printing debugging information so it probably wouldn't be of much help in your case.

**Reply by jschwartzenberg** [reply]
Is there a way to get a virtual serial port? I can choose the option, but then only `/dev/ttyS?` devices still are listed. (I'm using the AppImage.) Having something I could use `socat` on would be great!

**Comment by degaart** [comment]
Hello, can you elaborate on how to do that? Do I have to create a named pipe and redirect 86box's serial output to that pipe? Also, 86Box's configuration interface only lists bluetooth devices (the host does not have any real serial ports) so how would one redirect serial output to a named pipe? By manually editing the config file?

**Comment by OBattler** [comment]
I'm converting this to a discussion as this has to be (and is currently being, from what I see above) discussed first.

**Comment by raszpl** [comment]
Not a straight forward solution, but Putty logging to file works right now thru named pipe https://github.com/86Box/86Box/issues/257#issuecomment-1446123553 https://stackoverflow.com/questions/27477658/save-putty-output-to-file-from-command-line
